### ISSUE TITLE
关于论文的模型方法部分是不是有一些错误或者笔误

### ISSUE THREAD

**Issue by ShiMinghao0208** [wanggrun/SYSU-30k#4]
![image](https://user-images.githubusercontent.com/53176521/102297883-03b30500-3f8b-11eb-8da5-d169fe7b4641.png)
公式（4）是不是少了一个负号，否则最小化（4）式的含义不就是贴相同标签就可以得到最小值吗？
![image](https://user-images.githubusercontent.com/53176521/102297962-3957ee00-3f8b-11eb-948a-07c9fce3284d.png)
公式（6）应该是argmin吧，上面不是说minimize（2）式吗？

**Comment by wanggrun** [comment]
非常感谢您的仔细阅读。

1. 公式（4）是不是少了一个负号，否则最小化（4）式的含义不就是贴相同标签就可以得到最小值吗？答：并非如此。公式（4）没有少负号。请参考一下CRF的相关论文。直观上，由于公式（5）的存在，二元项是不希望打“不同标签”的，一旦打了“不同标签”，必然出现二元项损失（打“相同标签”没有二元损失）。当外观非常不相似时，二元项损失较小，此时才勉强接受打“不同标签”。

2. 标红一。答：（x_i, x_i）为（x_i, x_j）.

3. 标红二。答：y_i为y_j。

4. 公式（6）应该是argmin吧，上面不是说minimize（2）式吗？答：是的。非常感谢您的细心阅读。

**Comment by ShiMinghao0208** [comment]
> 非常感谢您的仔细阅读。
> 
> 1. 公式（4）是不是少了一个负号，否则最小化（4）式的含义不就是贴相同标签就可以得到最小值吗？答：并非如此。公式（4）没有少负号。请参考一下CRF的相关论文。直观上，由于公式（5）的存在，二元项是不希望打“不同标签”的，一旦打了“不同标签”，必然出现二元项损失（打“相同标签”没有二元损失）。当外观非常不相似时，二元项损失较小，此时才勉强接受打“不同标签”。
> 2. 标红一。答：（x_i, x_i）为（x_i, x_j）.
> 3. 标红二。答：y_i为y_j。
> 4. 公式（6）应该是argmin吧，上面不是说minimize（2）式吗？答：是的。非常感谢您的细心阅读。

好的，十分感谢您的回复！

**Comment by ShiMinghao0208** [comment]
请问参数中的batchid以及batchimage是什么意思呢，看代码不太能理解这两个参数的含义

**Comment by wanggrun** [comment]
请参考option.py里面的内容。

具体地，--batchid是指每个批量读入多少个人物ID;  --batchimage是指每个批量内每个ID读入多少张图。

谢谢。

**Comment by ShiMinghao0208** [comment]
> 请参考option.py里面的内容。
> 
> 具体地，--batchid是指每个批量读入多少个人物ID; --batchimage是指每个批量内每个ID读入多少张图。
> 
> 谢谢。

十分感谢解答，明白了

**Comment by Sixty-seven** [comment]
您好，1. 请问公式（6）应该改为argmin，那么相应代码部分CrossEntropyLoss.py中(31行)也需要将 targets = torch.argmax(targets * input_softmax, -1) 改为 targets = torch.argmin(targets * input_softmax, -1) ?
2. 我直接按照demo复现您的代码没问题，在market上为什么只要当bagid=1,2时能够复现出结果，而改为4或者其它参数(即一个bag中放更多id)就不能够跑出结果了？ 是不是还需要修改别的地方等？

期待您的回答，谢谢。

**Comment by ShiMinghao0208** [comment]
> 您好，1. 请问公式（6）应该改为argmin，那么相应代码部分CrossEntropyLoss.py中(31行)也需要将 targets = torch.argmax(targets * input_softmax, -1) 改为 targets = torch.argmin(targets * input_softmax, -1) ?
> 2. 我直接按照demo复现您的代码没问题，在market上为什么只要当bagid=1,2时能够复现出结果，而改为4或者其它参数(即一个bag中放更多id)就不能够跑出结果了？ 是不是还需要修改别的地方等？
> 
> 期待您的回答，谢谢。

他这里的argmax对应的不是（6） ，是（8）式。argmin是表示（6）的优化目标

**Comment by Sixty-seven** [comment]
> > 您好，1. 请问公式（6）应该改为argmin，那么相应代码部分CrossEntropyLoss.py中(31行)也需要将 targets = torch.argmax(targets * input_softmax, -1) 改为 targets = torch.argmin(targets * input_softmax, -1) ?
> > 2. 我直接按照demo复现您的代码没问题，在market上为什么只要当bagid=1,2时能够复现出结果，而改为4或者其它参数(即一个bag中放更多id)就不能够跑出结果了？ 是不是还需要修改别的地方等？
> > 期待您的回答，谢谢。
> 
> 他这里的argmax对应的不是（6） ，是（8）式。argmin是表示（6）的优化目标

好的，感谢解答。

**Comment by ShiMinghao0208** [comment]
请问文章中的图损失有一元项和二元项，我在代码CrossEntropyLoss那里看到了关于二元项的计算，但是好像没有一元项的计算啊，做弱监督训练的时候不需要一元项吗？还有这里的低级特征的比较是怎样比较的呢，看代码感觉不太像做了类似高斯核的运算啊。
不好意思，多多打扰了

**Comment by ShiMinghao0208** [comment]
请问关于数据集，一个bag里面类的分布您有相关统计吗，会出现一个bag中有的类数量极端的少有的类数量极端多的这种分布情形吗

**Comment by wanggrun** [comment]
感谢您的关注！

请问文章中的图损失有一元项和二元项，我在代码CrossEntropyLoss那里看到了关于二元项的计算，但是好像没有一元项的计算啊，做弱监督训练的时候不需要一元项吗？
答：需要一元项。在UnaryTerm_PairwiseTerm_WeaklyTripletLoss/loss/CrossEntropyLoss.py中，loss = self.cse_loss(inputs, targets) + 2. * loss_pairwise。其中self.cse_loss(inputs, targets)属于一元项。此时，target来自targets = torch.argmax(targets * input_softmax, -1)。

还有这里的低级特征的比较是怎样比较的呢，看代码感觉不太像做了类似高斯核的运算啊。
答：高斯核体现在这里：loss_pairwise = dist * inner_prod * mask，其中：
dist = torch.pow(low_level_feat, 2).sum(dim=1, keepdim=True).expand(n, n)
dist = dist + dist.t()
dist.addmm_(1, -2, low_level_feat, low_level_feat.t())
sigma = 1.
dist = torch.exp(- dist / sigma)

请问关于数据集，一个bag里面类的分布您有相关统计吗，会出现一个bag中有的类数量极端的少有的类数量极端多的这种分布情形吗
答：一个bag里面，类数量的数量可能为1，3，5这三种。

再次感谢您的关注！

**Comment by ShiMinghao0208** [comment]
> 感谢您的关注！
> 
> 请问文章中的图损失有一元项和二元项，我在代码CrossEntropyLoss那里看到了关于二元项的计算，但是好像没有一元项的计算啊，做弱监督训练的时候不需要一元项吗？
> 答：需要一元项。在UnaryTerm_PairwiseTerm_WeaklyTripletLoss/loss/CrossEntropyLoss.py中，loss = self.cse_loss(inputs, targets) + 2. * loss_pairwise。其中self.cse_loss(inputs, targets)属于一元项。此时，target来自targets = torch.argmax(targets * input_softmax, -1)。
> 
> 还有这里的低级特征的比较是怎样比较的呢，看代码感觉不太像做了类似高斯核的运算啊。
> 答：高斯核体现在这里：loss_pairwise = dist * inner_prod * mask，其中：
> dist = torch.pow(low_level_feat, 2).sum(dim=1, keepdim=True).expand(n, n)
> dist = dist + dist.t()
> dist.addmm_(1, -2, low_level_feat, low_level_feat.t())
> sigma = 1.
> dist = torch.exp(- dist / sigma)
> 
> 请问关于数据集，一个bag里面类的分布您有相关统计吗，会出现一个bag中有的类数量极端的少有的类数量极端多的这种分布情形吗
> 答：一个bag里面，类数量的数量可能为1，3，5这三种。
> 
> 再次感谢您的关注！
感谢您的回复，关于第三个问题，我的意思是一个bag里面假设有3类，那么三类的数目是比较平均的还是说有可能某个id的样本非常少，某个会很多这种比较极端的分布。我下载了数据集对照了一下label文件，不太明白，您这里bag级别的标签是按1，3，5，1，3，5这样来的，那么一个bag只有一个label那不就是全监督样本了吗，而且我看了一下bag里面应该也不止一个类啊，不太明白您这里的数据集怎么做的。希望得到您的答复，谢谢

**Comment by WongChen** [comment]
> 感谢您的关注！
> 
> 请问文章中的图损失有一元项和二元项，我在代码CrossEntropyLoss那里看到了关于二元项的计算，但是好像没有一元项的计算啊，做弱监督训练的时候不需要一元项吗？
> 答：需要一元项。在UnaryTerm_PairwiseTerm_WeaklyTripletLoss/loss/CrossEntropyLoss.py中，loss = self.cse_loss(inputs, targets) + 2. * loss_pairwise。其中self.cse_loss(inputs, targets)属于一元项。此时，target来自targets = torch.argmax(targets * input_softmax, -1)。
> 
> 还有这里的低级特征的比较是怎样比较的呢，看代码感觉不太像做了类似高斯核的运算啊。
> 答：高斯核体现在这里：loss_pairwise = dist * inner_prod * mask，其中：
> dist = torch.pow(low_level_feat, 2).sum(dim=1, keepdim=True).expand(n, n)
> dist = dist + dist.t()
> dist.addmm_(1, -2, low_level_feat, low_level_feat.t())
> sigma = 1.
> dist = torch.exp(- dist / sigma)
> 
> 请问关于数据集，一个bag里面类的分布您有相关统计吗，会出现一个bag中有的类数量极端的少有的类数量极端多的这种分布情形吗
> 答：一个bag里面，类数量的数量可能为1，3，5这三种。
> 
> 再次感谢您的关注！

https://github.com/wanggrun/SYSU-30k/blob/master/GraphReID/UnaryTerm_PairwiseTerm_WeaklyTripletLoss/loss/CrossEntropyLoss.py#L47
这行代码和公式8 的二元项不太一样？
公式8 是： 
![image](https://user-images.githubusercontent.com/10301728/104807490-631a6400-581a-11eb-899f-2108305fb724.png)
代码中的是：  {P_i}^T*log(P_i)
理解错了的话还望指正

**Comment by weleen** [comment]
> > 感谢您的关注！
> > 请问文章中的图损失有一元项和二元项，我在代码CrossEntropyLoss那里看到了关于二元项的计算，但是好像没有一元项的计算啊，做弱监督训练的时候不需要一元项吗？
> > 答：需要一元项。在UnaryTerm_PairwiseTerm_WeaklyTripletLoss/loss/CrossEntropyLoss.py中，loss = self.cse_loss(inputs, targets) + 2. * loss_pairwise。其中self.cse_loss(inputs, targets)属于一元项。此时，target来自targets = torch.argmax(targets * input_softmax, -1)。
> > 还有这里的低级特征的比较是怎样比较的呢，看代码感觉不太像做了类似高斯核的运算啊。
> > 答：高斯核体现在这里：loss_pairwise = dist * inner_prod * mask，其中：
> > dist = torch.pow(low_level_feat, 2).sum(dim=1, keepdim=True).expand(n, n)
> > dist = dist + dist.t()
> > dist.addmm_(1, -2, low_level_feat, low_level_feat.t())
> > sigma = 1.
> > dist = torch.exp(- dist / sigma)
> > 请问关于数据集，一个bag里面类的分布您有相关统计吗，会出现一个bag中有的类数量极端的少有的类数量极端多的这种分布情形吗
> > 答：一个bag里面，类数量的数量可能为1，3，5这三种。
> > 再次感谢您的关注！
> 
> https://github.com/wanggrun/SYSU-30k/blob/master/GraphReID/UnaryTerm_PairwiseTerm_WeaklyTripletLoss/loss/CrossEntropyLoss.py#L47
> 这行代码和公式8 的二元项不太一样？
> 公式8 是：
> ![image](https://user-images.githubusercontent.com/10301728/104807490-631a6400-581a-11eb-899f-2108305fb724.png)
> 代码中的是： {P_i}^T*log(P_i)
> 理解错了的话还望指正

代码里面的二元项是{P_i}^T{P_j}吧
